### ISSUE TITLE
Oriented Bounding Box Tracker

### ISSUE THREAD

**Issue by LilBabines** [mikel-brostrom/boxmot#1774]
### Search before asking

- [X] I have searched the Yolo Tracking [issues](https://github.com/mikel-brostrom/yolo_tracking/issues) and found no similar bug report.

### Question

Hello,

I’m currently working on a project that requires object tracking with **Oriented Bounding Boxes (OBB)**. Despite thorough research, I haven’t found any convincing implementation for a tracker specifically handling OBB, either in this repository or elsewhere. For instance, the `yolo track model=yolov11m-obb` command and the code from [ultralytics/trackers](https://github.com/ultralytics/ultralytics/tree/main/ultralytics/trackers) don’t seem to clearly integrate OBB. 

In this regard, I’d like to ask:

- Do you think it would be beneficial to adapt a tracker to support OBB, for example, by adding a dimension to the KalmannFilter  for angle prediction ?
- Alternatively, would using [centroid](https://github.com/mikel-brostrom/boxmot/blob/20363a40db08b1b0effbe427e398809e4c607969/boxmot/utils/iou.py#L132) association function or adapting [iou_batch](https://github.com/mikel-brostrom/boxmot/blob/20363a40db08b1b0effbe427e398809e4c607969/boxmot/utils/iou.py#L20C4-L36C17) be sufficient ?
- Or would it be better to avoid the problem altogether with something like : `update(dets = obb_to_x1y1x2y2(boxes))`

Thank you in advance for your support and for the excellent work on this package! I’m looking forward to your feedback.

Best regards,

**Comment by mikel-brostrom** [comment]
> Do you think it would be beneficial to adapt a tracker to support OBB, for example, by adding a dimension to the Kalman Filter for angle prediction ?

"Straightening" up the bounding boxes will give you worse association with any type of IoU. Think about the case of a top down view where two cars are side to side. With OBB there would be zero overlap but there may be if they are straightened. If your use-case is simple, it may be possible to simplify the problem by a `obb_to_x1y1x2y2` function.

**Comment by LilBabines** [comment]
Thank you very much for your quick response. As you mentioned, "straightening" up the bounding boxes is indeed risky, and I fully understand the challenges related to oriented versus non-oriented bounding boxes.

I’ve attached an image to illustrate my use case (quite similar to #1725), which I believe is very well-suited for oriented bounding boxes. I’m considering starting with **OCSort** for a more or less complete adaptation to handle oriented bounding boxes. I will also take a closer look later to consider feature extraction if needed.

Do you think this could be a valuable enhancement and interest the community?

![frame_obb](https://github.com/user-attachments/assets/c52c4e9a-5a58-45ef-9c06-b1da224afbd5)

**Comment by mikel-brostrom** [comment]
> Do you think this could be a valuable enhancement and interest the community?

Absolutely. OBB tracking would be a natural next step for this repo as well given the wider adoption of these object detection methods. I recommend you to trigger the load of a specific Kalman Filter designed for OBB when the detection input is one of (`xyxyxyxy`, `xywhr`) or whatever your OBB format is.

**Comment by mikel-brostrom** [comment]
OBB straightening:

```py
import cv2
import math
import numpy as np

def extract_and_straighten_crop(image, cx, cy, w, h, r):
    """
    Extract and straighten a crop from an image given an OBB.

    Parameters
    ----------
    image : numpy.ndarray
        Input image (BGR or RGB).
    cx, cy : float
        Center coordinates of the OBB.
    w, h : float
        Width and height of the OBB.
    r : float
        Rotation angle in radians.
        The OBB is rotated by r around (cx, cy).
    
    Returns
    -------
    crop : numpy.ndarray
        The extracted, straightened crop.
    """

    # Convert rotation angle from radians to degrees
    angle_degrees = math.degrees(r)

    # Compute rotation matrix
    # We rotate the entire image by -angle_degrees to align the OBB with axes
    M = cv2.getRotationMatrix2D((cx, cy), -angle_degrees, 1.0)

    # Determine the size of the image
    rows, cols = image.shape[:2]

    # Rotate the entire image so that the OBB is axis-aligned
    rotated = cv2.warpAffine(image, M, (cols, rows), flags=cv2.INTER_LINEAR, borderMode=cv2.BORDER_CONSTANT, borderValue=(122, 122, 122))

    # Now, after rotation, the bounding box corresponding to the OBB is axis-aligned.
    # The bounding box corners in the rotated image:
    x_min = int(cx - w/2)
    y_min = int(cy - h/2)
    x_max = int(cx + w/2)
    y_max = int(cy + h/2)

    # Extract the crop from the rotated image
    # We must ensure the indices are within image bounds. Add safety clipping.
    x_min = max(0, x_min)
    y_min = max(0, y_min)
    x_max = min(cols, x_max)
    y_max = min(rows, y_max)

    crop = rotated[y_min:y_max, x_min:x_max]

    return crop

# Example usage
if __name__ == "__main__":
    # Load an image
    image = cv2.imread("bus.jpg")

    # Suppose you have an OBB: center (100, 200), width=50, height=30, rotated by 45 degrees.
    cx, cy = 740, 636
    w, h = 138, 483
    r_degrees = 45
    r_radians = math.radians(r_degrees)

    straight_crop = extract_and_straighten_crop(image, cx, cy, w, h, r_radians)

    # Save or display the result
    cv2.imshow("Straightened Crop", straight_crop)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
```

Could be used for feature extraction later on 🚀

**Comment by LilBabines** [comment]
Hello, 
I’m almost done with an OCSORT tracker for oriented bounding boxes. You can find my contribution in a BoxMot fork at `master-obb` branch : [LilBabines/boxmot](https://github.com/LilBabines/boxmot/tree/master-obb). I’ll make a pull request as soon as possible. 🚀
I still have a few things to finalize, such as adding a new association function `batch_iou_obb` ( probaly using `cv2.rotatedRectangleIntersection()` ), or adapting the camera motion constraint. My code isn’t very clean yet; I’ve used an argument `tracker.is_off` to adjust the behavior of functions affected by rotation. There are certainly more elegant solutions, like using decorators (similar to how per-class-based tracking is handled). 
I would be happy to receive any feedback, whether positive or negative, on my addition. Feel free to reach out if you have any questions ! 
Thank you, and see you soon!

**Comment by mikel-brostrom** [comment]
Awesome @LilBabines ! Looking forward to this 😄

**Comment by mikel-brostrom** [comment]
Will find some time next week for feedback of what is implemented ☺️
